From 0522a3f3d26387530088b834b9e5212a9dce9588 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Wed, 08 Dec 2021 08:23:43 +0200
Subject: [PATCH] [SW-63186] gaudi2: security: add axprot base infra

Axprot values for various interfaces needs to be updated in preboot.
These interfaces consist of multiple RTRs over multiple dcores.

For eg: NIC has 12 instances with 2 rtrs for each instance, and MME
sbte has 4 dcores and 5 rtr for each instance respectively.

Hence adding generic function to support axprot programming in preboot.

Change-Id: I5ad82f68ff1b6a25c464d23e12499edc627a7e2f
---

diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index 276fcf9..e558a33 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -891,6 +891,29 @@
 	hl_secure_hbw_rr(HL_SEC_HBW_HBM, 0, 0, rr_sel);
 }
 
+/* Reg format is <DCORE_NUM>_RTR<INST_NUM>_MSTR_IF */
+struct axprot_regs {
+	u32 base_reg;
+	u32 value;
+	int dcore_dist;
+	int inst_dist;
+	u8 dcore_num;
+	u8 inst_num;
+};
+
+__unused static void preboot_axprot_regs_init(struct axprot_regs *reg)
+{
+	int i, j, dcore_reg, rtr_reg;
+
+	for (i = 0; i < reg->dcore_num; i++) {
+		dcore_reg = reg->base_reg + i * reg->dcore_dist;
+		for (j = 0; j < reg->inst_num; j++) {
+			rtr_reg = dcore_reg + j * reg->inst_dist;
+			regmap_writel(reg->value, rtr_reg);
+		}
+	}
+}
+
 void preboot_security_init(void)
 {
 	preboot_secure_pcie();
